### PR TITLE
Add brimstone chest support to loot tracker

### LINKED ISSUE #7749: Brimstone key loot in loot tracker
**Is your feature request related to a problem? Please describe.**
The Brimstone keys don't appear in the loot tracker

**Describe the solution you'd like**
Make entries for each chest opened with the key.

### PR DESCRIPTION
**Pull request by AdamBallinger**
Adds loot from brimstone chest to the loot tracker.

789b2005384d3502525817c6f97a05d2

Closes #7749

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 1]
```diff
@@ -120,6 +130,8 @@
 	private LootTrackerPanel panel;
 	private NavigationButton navButton;
 	private String eventType;
+	// Brimstone chest loot message
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: Name the variable better so this comment isn't needed. `brimstoneChestMessage` maybe?

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 2]
```diff
@@ -93,6 +95,14 @@
 	private static final Pattern CLUE_SCROLL_PATTERN = Pattern.compile("You have completed [0-9]+ ([a-z]+) Treasure Trails.");
 	private static final int THEATRE_OF_BLOOD_REGION = 12867;
 
+	// Brimstone loot handling
+	private final String BRIMSTONE_CHEST_MESSAGE = "You find some treasure in the chest!";
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: All of these can be `static`.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 3]
```diff
@@ -93,6 +95,14 @@
...
+	private final String BRIMSTONE_CHEST_EVENT_TYPE = "Brimstone Chest";
+	private final Pattern VALUABLE_DROP_PATTERN = Pattern.compile("Valuable drop: ([0-9,]+) x ([-A-z-0-9 ]+)");
+	private final Pattern VALUABLE_DROP_PATTERN2 = Pattern.compile("Valuable drop: ([-A-z-0-9 ]+)");
+	private final Pattern PLAYER_RECEIVED_DROP_PATTERN = Pattern.compile("([A-z0-9 ]+) received a drop: ([0-9,]+) x ([-A-z0-9 ]+)");
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: Don't capture player name in a group if it won't be used.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 4]
```diff
@@ -93,6 +95,14 @@
...
+	private final Pattern PLAYER_RECEIVED_DROP_PATTERN2 = Pattern.compile("([A-z0-9 ]+) received a drop: ([-A-z0-9 ]+)");
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: Don't capture player name in a group if it won't be used.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 5]
```diff
@@ -389,6 +413,102 @@ public void onChatMessage(ChatMessage event)
...
+
+		if (lootTrackerClient != null && config.saveLoot())
+		{
+			LootRecord lootRecord = new LootRecord(eventType, LootRecordType.EVENT, toGameItems(items), Instant.now());
+			lootTrackerClient.submit(lootRecord);
+		}
+	}
+
+	/*
+	Searches for and returns an item in the provided item container
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]:
```suggestion
	 * Searches for and returns an item in the provided item container
```

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 6]
```diff
@@ -389,6 +413,102 @@ public void onChatMessage(ChatMessage event)
...
+			itemQuantity = 1;
+			itemName = matcher.group(1).trim();
+
+			submitBrimstoneChestLoot(itemName, itemQuantity, container);
+		}
+	}
+
+	private void submitBrimstoneChestLoot(String itemName, int itemQuantity, ItemContainer container)
+	{
+		final Collection<ItemStack> items = Arrays.stream(new Item[]{getItem(itemName, container)})
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: This can just use item manager?

**Comment by fargoldbee** [comment]
Closes #7749 if done

Not sure if this is still being developped?

**Comment by Adam-** [comment]
Did you considering copying the inventory at the first message and diffing them? Thats what the mlm plugin ore counter does this works well.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 7]
```diff
@@ -93,6 +95,14 @@
 	private static final Pattern CLUE_SCROLL_PATTERN = Pattern.compile("You have completed [0-9]+ ([a-z]+) Treasure Trails.");
 	private static final int THEATRE_OF_BLOOD_REGION = 12867;
 
+	// Brimstone loot handling
+	private static final String BRIMSTONE_CHEST_MESSAGE = "You find some treasure in the chest!";
+	private static final String BRIMSTONE_CHEST_EVENT_TYPE = "Brimstone Chest";
+	private static final Pattern VALUABLE_DROP_PATTERN = Pattern.compile("Valuable drop: ([0-9,]+) x ([-A-z-0-9 ]+)");
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: Both of these can be one pattern: `Valuable drop: (?:([0-9,]+) x )?([-A-z-0-9 ]+)`

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 8]
```diff
@@ -389,6 +412,90 @@ public void onChatMessage(ChatMessage event)
...
+		// Coins are not included in the item price search so manually add it.
+		if (itemName.equals("Coins"))
+		{
+			final ItemPrice coinsPrice = new ItemPrice();
+			coinsPrice.setId(ItemID.COINS_995);
+			coinsPrice.setName(itemName);
+			itemPrices.add(coinsPrice);
+		}
+
+		final Collection<ItemStack> items = Arrays.stream(itemPrices.toArray())
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: This is just `itemPrices.toArray().stream()`

**Comment by AdamBallinger** [comment]
I made the changes to diff check the inventory like you suggested which definitely works better, thank you.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 9]
```diff
@@ -389,6 +414,51 @@ public void onChatMessage(ChatMessage event)
...
+	{
+		if (inventoryContainer != null && inventorySnapshot != null)
+		{
+			Multiset<Integer> currentInventory = HashMultiset.create();
+			Arrays.stream(inventoryContainer.getItems())
+					.forEach(item -> currentInventory.add(item.getId(), item.getQuantity()));
+
+			final Multiset<Integer> diff = Multisets.difference(currentInventory, inventorySnapshot);
+
+			diff.forEachEntry(this::submitBrimstoneChestLoot);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: This submits one loot entry per item... is there always just one item?

Comment (15987632) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: according to the wiki everything dropped is either noted, stackable, or dropped in a quantity of 1. So yes

**Pull request merged**